### PR TITLE
Buildkite QA

### PR DESCRIPTION
**Pull request by matt-chan**
testing

### CODE REVIEW

**Comment by matt-chan** [comment]
Mergeable now. I've gotten it down to 5 minutes or so. I can probably make it to 2 or 3 minutes, but I need more cores. I'm getting resource contention I think.

As far as I can see, it should perform the exact same tests as Travis, but more complete (Ubuntu, Fedora 23/24) and also faster.

**Comment by matt-chan** [comment]
Down to 3m. Woot =). 

I'm going to have a schedule the regression tests nightly or something though, at least until I get the rest  of the build agents online in Hamilton.

File: tools/qa/buildkite_static_code_analysis.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    trapdoor_cpplint.py
+    trapdoor_doxygen.py
+    trapdoor_import.py
+    trapdoor_pycodestyle.py
+    trapdoor_pydocstyle.py"
+
+    for i in ${TRAPDOORS}; do
+        tools/qa/simulate_trapdoor_pr.py -vA $SHA tools/qa/$i
+    done
+fi
```
Comment (tovrstra) [inline tools/qa/buildkite_static_code_analysis.sh]: EOL missing

File: tools/qa/buildkite_common.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        API_URL=`echo "$BUILDKITE_REPO" | sed "s/git@/https:\/\/api./" | sed "s/\.com:/\.com\/repos\//" | sed "s/\.git$/\/pulls\/$BUILDKITE_PULL_REQUEST/"`
+        MERGESTATE="null"
+
+        while [ "$MERGESTATE" = "null" ]; do
+            MERGESTATE=`curl $API_URL | jq .mergeable`
+            if [ "$MERGESTATE" = "true" ]; then
+                git fetch -f origin pull/$BUILDKITE_PULL_REQUEST/merge:temp_merge
+                git checkout temp_merge
+            elif [ "$MERGESTATE" = "null" ]; then
+                # git github a chance to compute mergeability
```
Comment (tovrstra) [inline tools/qa/buildkite_common.sh]: s/# git/# give/

**Comment by tovrstra** [comment]
Looks good! Other than the silly things this is good to go. Thanks!

File: tools/qa/buildkite_static_code_analysis.sh (hunk 1 above)
Comment (matt-chan) [inline tools/qa/buildkite_static_code_analysis.sh]: It complains about a whitespace error if I add a newline at the end?

Comment (tovrstra) [inline tools/qa/buildkite_static_code_analysis.sh]: not sure what's going on. I get this weird symbol on the github web page that something is wrong with the line ending. Just a sec...

Comment (tovrstra) [inline tools/qa/buildkite_static_code_analysis.sh]: So this is it: there is no newline at the end of the file. You don't have to add a empty line. You just need to make sure that the last line ends with a new-line character, which is the conventional thing to do. Try the following:

    toon@molmod55 ~/univ/code/horton1:buildkiteQA> cat tools/qa/buildkite_code_analysis.sh 
    #!/usr/bin/env bash

    if [ "$BUILDKITE_PULL_REQUEST" != "false" ]; then
        source tools/qa/buildkite_common.sh
        checkout_merge_commit
        get_ancestor #writes $SHA variable

        echo "--- Build refatoms"
        rm -rf data/refatoms/*.h5 data/refatoms/*.tar.bz2
        make -C data/refatoms/

        echo "--- Running trapdoors tests"
        ./tools/qa/check_whitespace.py || report_error "Whitespace errors in some commits"

        TRAPDOORS="trapdoor_coverage.py
        trapdoor_namespace.py"

        export PYTHONPATH=`pwd`
        export HORTONDATA=`pwd`/data

        for i in ${TRAPDOORS}; do
            tools/qa/simulate_trapdoor_pr.py -vrA $SHA tools/qa/$i
        done
    fitoon@molmod55 ~/univ/code/horton1:buildkiteQA> 

Note the `fi` being on the same line as my command line. It is unfortunate that `git diff --check` does allow for this.

Comment (tovrstra) [inline tools/qa/buildkite_static_code_analysis.sh]: See http://stackoverflow.com/questions/5813311/no-newline-at-end-of-file and https://robots.thoughtbot.com/no-newline-at-end-of-file for some background.

I'll see what we can do to check this in pre-commit.

Comment (matt-chan) [inline tools/qa/buildkite_static_code_analysis.sh]: Ah I see it. Okay this should fix it.

**Comment by matt-chan** [comment]
Okay, you're good to go @tovrstra. Sorry for the delay! I forgot to check on the page again =S

**Comment by tovrstra** [comment]
No problem. Thanks!

**Pull request merged**